### DISCUSSION TITLE
Use one operator to manage multiple CRDs

### DISCUSSION THREAD

**Discussion by gajananbhat** [nolar/kopf#814] [General]
How do I use one operator to manage multiple crds?

```
@kopf.on.create('r1.com', 'v1', 'power')
@kopf.on.create('r1.com', 'v1', 'temp')
def create_fn(meta, spec, namespace, logger, body, kind, **kwargs):
```

The above doesn't work.

Also, how do I get the kind of Custom Resource inside the create_fn? "kind" argument does not work.

Thanks

**Comment by nolar** [comment]
It should work, you are doing it right. Can you please explain how exactly that does not work? Specifically, the setup and code snippets (CRD yamls) so that I could reproduce the case where it does not work.

To know the resource, there is a kwarg named `resource`. This is a dataclass with fields representing all known information about the resource (as taken from the initial cluster scanning). So, `resource.kind` can be used to get the kind.

* https://kopf.readthedocs.io/en/stable/kwargs/#resource-related-kwargs
* https://kopf.readthedocs.io/en/stable/packages/kopf/#kopf.Resource

Also, you can use `param` to distinguish several decorators/causes in the same function.

* https://kopf.readthedocs.io/en/stable/kwargs/#parametrization

**Comment by gajananbhat** [comment]
Hi @nolar, 

Apologies for the delay. Yes. It works now. It was a typo on my part that caused the confusion. 

Also, thanks for the pointer to "resource.kind". I am now able to process CRs based on their kind.

How can we return messages to kubectl output in case of any failure during creation/deletion?

Thanks

**Reply by nolar** [reply]
There is no way to return messages to kubectl with create/update/delete handlers. They are invited after everything has happened. 

But you can take a look at admission hooks, which can either reject the creation/update (with a message), or accept/modify the object and yield some warnings to the user.

**Reply by gajananbhat** [reply]
Thank you. Is there a way to integrate custom resource logs with "kubectl logs" output? I can see the output if I do "kubectl describe" on a custom resource. But, it would be great if we can get it with kubectl logs.
